File: src/main/java/com/chm/plugin/idea/forestx/utils/TreeNodeUtil.java
package com.chm.plugin.idea.forestx.utils;

import com.chm.plugin.idea.forestx.Icons;
import com.chm.plugin.idea.forestx.annotation.Annotation;
import com.google.common.collect.Maps;
import com.intellij.codeInsight.AnnotationUtil;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.psi.PsiAnnotation;
import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiMethod;
import com.intellij.util.ui.tree.TreeUtil;
import org.apache.commons.compress.utils.Lists;

import javax.swing.*;
import javax.swing.tree.DefaultMutableTreeNode;
import javax.swing.tree.DefaultTreeModel;
import javax.swing.tree.MutableTreeNode;
import javax.swing.tree.TreeNode;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Objects;

/**
 * @author caihongming
 * @version v1.0
 * @since 2022-08-02
 **/
public class TreeNodeUtil {

    private static final Map<String, Icon> METHOD_ICON_MAP = Maps.newConcurrentMap();

    private static final Map<String, Icon> REQUEST_TYPE_ICON_MAP = Maps.newConcurrentMap();

    public static final Comparator NAME_SORTED = (o1, o2) -> {
        String name1 = TreeNodeUtil.getNodeName(o1);
        String name2 = TreeNodeUtil.getNodeName(o2);
        return name1.compareTo(name2);
    };

    static {
        METHOD_ICON_MAP.put(Annotation.GET.getQualifiedName(), Icons.GET);
        METHOD_ICON_MAP.put(Annotation.GET_REQUEST.getQualifiedName(), Icons.GET);
        METHOD_ICON_MAP.put(Annotation.POST.getQualifiedName(), Icons.POST);
        METHOD_ICON_MAP.put(Annotation.POST_REQUEST.getQualifiedName(), Icons.POST);
        METHOD_ICON_MAP.put(Annotation.PUT.getQualifiedName(), Icons.PUT);
        METHOD_ICON_MAP.put(Annotation.PUT_REQUEST.getQualifiedName(), Icons.PUT);
        METHOD_ICON_MAP.put(Annotation.PATCH.getQualifiedName(), Icons.PATCH);
        METHOD_ICON_MAP.put(Annotation.PATCH_REQUEST.getQualifiedName(), Icons.PATCH);
        METHOD_ICON_MAP.put(Annotation.HEAD_REQUEST.getQualifiedName(), Icons.HEAD);
        METHOD_ICON_MAP.put(Annotation.OPTIONS.getQualifiedName(), Icons.OPTIONS);
        METHOD_ICON_MAP.put(Annotation.OPTIONS_REQUEST.getQualifiedName(), Icons.OPTIONS);
        METHOD_ICON_MAP.put(Annotation.DELETE.getQualifiedName(), Icons.DELETE);
        METHOD_ICON_MAP.put(Annotation.DELETE_REQUEST.getQualifiedName(), Icons.DELETE);
        METHOD_ICON_MAP.put(Annotation.TRACE.getQualifiedName(), Icons.TRACE);
        METHOD_ICON_MAP.put(Annotation.TRACE_REQUEST.getQualifiedName(), Icons.TRACE);

        REQUEST_TYPE_ICON_MAP.put("GET", Icons.GET);
        REQUEST_TYPE_ICON_MAP.put("POST", Icons.POST);
        REQUEST_TYPE_ICON_MAP.put("PUT", Icons.PUT);
        REQUEST_TYPE_ICON_MAP.put("HEAD", Icons.HEAD);
        REQUEST_TYPE_ICON_MAP.put("OPTIONS", Icons.OPTIONS);
        REQUEST_TYPE_ICON_MAP.put("DELETE", Icons.DELETE);
        REQUEST_TYPE_ICON_MAP.put("PATCH", Icons.PATCH);
        REQUEST_TYPE_ICON_MAP.put("TRACE", Icons.TRACE);
    }

    public static DefaultMutableTreeNode findNode(TreeNode parentNode, Object o) {
        for (int i = 0; i < parentNode.getChildCount(); i++) {
            TreeNode child = parentNode.getChildAt(i);
            String childName = getNodeName(child);
            String oName = getNodeName(o);
            if (Objects.equals(childName, oName)) {
                return (DefaultMutableTreeNode) child;
            }
        }
        return null;
    }

    public static DefaultMutableTreeNode findNodeOrNew(DefaultTreeModel model,
                                                       DefaultMutableTreeNode parentNode,
                                                       Object o) {
        return findNodeOrNew(model, parentNode, o, false);
    }

    public static DefaultMutableTreeNode findNodeOrNew(DefaultTreeModel model,
                                                       DefaultMutableTreeNode parent,
                                                       Object o,
                                                       boolean sorted) {
        DefaultMutableTreeNode node = findNode(parent, o);
        if (node != null) {
            return node;
        }
        DefaultMutableTreeNode newNode = new DefaultMutableTreeNode(o);
        if (sorted) {
            TreeUtil.insertNode(newNode, parent, model, NAME_SORTED);
        } else {
            int insertionPoint = parent.getChildCount() >= 0 ? parent.getChildCount() : -(parent.getChildCount() + 1);
            if (model != null) {
                model.insertNodeInto(newNode, parent, insertionPoint);
            } else {
                parent.insert(newNode, insertionPoint);
            }
        }
        return newNode;
    }

    public static DefaultMutableTreeNode findNodeOrNew(DefaultTreeModel model,
                                                       DefaultMutableTreeNode parent,
                                                       Object o,
                                                       int index) {
        DefaultMutableTreeNode node = findNode(parent, o);
        if (node != null) {
            return node;
        }
        return insertNode(model, parent, o, index);
    }

    public static DefaultMutableTreeNode moveOrInsert(DefaultTreeModel model,
                                                      DefaultMutableTreeNode parent,
                                                      Object o,
                                                      int index) {
        DefaultMutableTreeNode node = findNode(parent, o);
        if (node != null && parent.getChildCount() < index) {
            TreeNode child = parent.getChildAt(index);
            if (child == node) {
                return node;
            }
            removeNode(model, parent, node);
            insertNode(model, parent, node, index);
            return node;
        }
        return insertNode(model, parent, o, index);
    }

    public static DefaultMutableTreeNode insertNode(DefaultTreeModel model,
                                                    DefaultMutableTreeNode parent,
                                                    Object o,
                                                    int index) {
        DefaultMutableTreeNode newNode = new DefaultMutableTreeNode(o);
        int insertionPoint = index >= 0 ? index : -(index + 1);
        if (model != null) {
            model.insertNodeInto(newNode, parent, insertionPoint);
        } else {
            parent.insert(newNode, insertionPoint);
        }
        return newNode;
    }

    public static void insertNode(DefaultTreeModel model,
                                  DefaultMutableTreeNode parent,
                                  DefaultMutableTreeNode child,
                                  int index) {
        int insertionPoint = index >= 0 ? index : -(index + 1);
        if (model != null) {
            model.insertNodeInto(child, parent, insertionPoint);
        } else {
            parent.insert(child, insertionPoint);
        }
    }

    public static DefaultMutableTreeNode addNode(DefaultTreeModel model,
                                                 DefaultMutableTreeNode parent,
                                                 Object o) {
        DefaultMutableTreeNode newNode = new DefaultMutableTreeNode(o);
        int insertionPoint = parent.getChildCount() >= 0 ? parent.getChildCount() : -(parent.getChildCount() + 1);
        if (model != null) {
            model.insertNodeInto(newNode, parent, insertionPoint);
        } else {
            parent.insert(newNode, insertionPoint);
        }
        return newNode;
    }

    public static List<MutableTreeNode> getAllChildren(DefaultMutableTreeNode parent) {
        List<MutableTreeNode> childrenList = Lists.newArrayList();
        for (int i = 0; i < parent.getChildCount(); i++) {
            MutableTreeNode child = (MutableTreeNode) parent.getChildAt(i);
            childrenList.add(child);
        }
        return childrenList;
    }


    public static void removeAllChildren(DefaultTreeModel model,
                                         DefaultMutableTreeNode parent) {
        for (int i = parent.getChildCount() - 1; i >= 0; i--) {
            MutableTreeNode child = (MutableTreeNode) parent.getChildAt(i);
            removeNode(model, parent, child);
        }

    }

    public static void removeChildrenList(DefaultTreeModel model,
                                          DefaultMutableTreeNode parent,
                                          List<MutableTreeNode> childrenList) {
        for (MutableTreeNode children : childrenList) {
            if (model != null) {
                model.removeNodeFromParent(children);
            } else {
                parent.remove(children);
            }
        }

    }

    public static void removeNode(DefaultTreeModel model,
                                  DefaultMutableTreeNode parent,
                                  MutableTreeNode child) {
        if (model != null) {
            model.removeNodeFromParent(child);
        } else {
            parent.remove(child);
        }
    }

    public static String getNodeName(Object node) {
        Object o;
        if (node instanceof DefaultMutableTreeNode) {
            o = ((DefaultMutableTreeNode) node).getUserObject();
        } else {
            o = node;
        }
        if (o instanceof Project) {
            return ((Project) o).getName();
        } else if (o instanceof Module) {
            return ((Module) o).getName();
        } else if (o instanceof PsiClass) {
            return ((PsiClass) o).getQualifiedName();
        } else if (o instanceof PsiMethod) {
            return ((PsiMethod) o).getName();
        } else {
            return o != null ? o.toString() : "null";
        }
    }

    public static Icon getNodeIcon(Object node) {
        Object o;
        if (node instanceof DefaultMutableTreeNode) {
            o = ((DefaultMutableTreeNode) node).getUserObject();
        } else {
            o = node;
        }
        if (o instanceof Project) {
            return Icons.PROJECT_16;
        } else if (o instanceof Module) {
            return Icons.MODULE_16;
        } else if (o instanceof PsiClass) {
            return Icons.INTERFACE_18;
        } else if (o instanceof PsiMethod) {
            PsiAnnotation[] methodAnnotations = AnnotationUtil.getAllAnnotations((PsiMethod) o, false, null);
            for (PsiAnnotation methodAnnotation : methodAnnotations) {
                Icon icon = METHOD_ICON_MAP.get(methodAnnotation.getQualifiedName());
                if (icon != null) {
                    return icon;
                }

                if (Objects.equals(methodAnnotation.getQualifiedName(), Annotation.REQUEST.getQualifiedName())) {
                    String type = AnnotationUtil.getStringAttributeValue(methodAnnotation, "type");
                    Icon icon2 = REQUEST_TYPE_ICON_MAP.get(StringUtil.toUpperCase(type));
                    if (icon2 != null) {
                        return icon2;
                    }
                    return Icons.GET;
                }
            }
            return Icons.GET;
        } else {
            return Icons.ICON_16;
        }
    }
}

File: src/main/kotlin/com/chm/plugin/idea/forestx/utils/TreeNodeUtil.kt
package com.chm.plugin.idea.forestx.utils

import com.chm.plugin.idea.forestx.Icons
import com.chm.plugin.idea.forestx.annotation.Annotation
import com.chm.plugin.idea.forestx.template.psi.ForestTemplatePathElement
import com.chm.plugin.idea.forestx.template.psi.ForestTemplatePrimary
import com.chm.plugin.idea.forestx.template.utils.ForestTemplateUtil
import com.intellij.codeInsight.AnnotationUtil
import com.intellij.icons.AllIcons
import com.intellij.openapi.module.Module
import com.intellij.openapi.project.Project
import com.intellij.openapi.roots.ProjectRootManager
import com.intellij.openapi.util.text.StringUtil
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.*
import com.intellij.psi.search.GlobalSearchScope
import com.intellij.psi.search.GlobalSearchScopesCore
import com.intellij.psi.search.ProjectScope
import com.intellij.psi.search.SearchScope
import com.intellij.psi.search.searches.AnnotationTargetsSearch
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.util.ui.tree.TreeUtil
import org.apache.commons.compress.utils.Lists
import org.jetbrains.yaml.YAMLUtil
import org.jetbrains.yaml.psi.YAMLFile
import org.jetbrains.yaml.psi.YAMLKeyValue
import org.jetbrains.yaml.psi.YAMLPsiElement
import org.jetbrains.yaml.psi.YAMLValue
import org.jetbrains.yaml.psi.impl.YAMLBlockMappingImpl
import org.jetbrains.yaml.psi.impl.YAMLBlockSequenceImpl
import java.util.function.BiFunction
import javax.swing.Icon
import javax.swing.tree.*

/**
 * @author caihongming
 * @version v1.0
 * @since 2022-08-25
 **/
private val METHOD_ICON_MAP: MutableMap<String, Icon> = mutableMapOf(
    Annotation.GET.qualifiedName to Icons.GET,
    Annotation.GET_REQUEST.qualifiedName to Icons.GET,
    Annotation.POST.qualifiedName to Icons.POST,
    Annotation.POST_REQUEST.qualifiedName to Icons.POST,
    Annotation.PUT.qualifiedName to Icons.PUT,
    Annotation.PUT_REQUEST.qualifiedName to Icons.PUT,
    Annotation.PATCH.qualifiedName to Icons.PATCH,
    Annotation.PATCH_REQUEST.qualifiedName to Icons.PATCH,
    Annotation.HEAD_REQUEST.qualifiedName to Icons.HEAD,
    Annotation.OPTIONS.qualifiedName to Icons.OPTIONS,
    Annotation.OPTIONS_REQUEST.qualifiedName to Icons.OPTIONS,
    Annotation.DELETE.qualifiedName to Icons.DELETE,
    Annotation.DELETE_REQUEST.qualifiedName to Icons.DELETE,
    Annotation.TRACE.qualifiedName to Icons.TRACE,
    Annotation.TRACE_REQUEST.qualifiedName to Icons.TRACE
)

private val REQUEST_TYPE_ICON_MAP: MutableMap<String, Icon> = mutableMapOf(
    "GET" to Icons.GET,
    "POST" to Icons.POST,
    "PUT" to Icons.PUT,
    "HEAD" to Icons.HEAD,
    "OPTIONS" to Icons.OPTIONS,
    "DELETE" to Icons.DELETE,
    "PATCH" to Icons.PATCH,
    "TRACE" to Icons.TRACE
)

val NAME_SORTED: Comparator<Any> = Comparator { o1: Any, o2: Any ->
    val name1 = o1.getNodeName()
    val name2 = o2.getNodeName()
    name1.compareTo(name2)
}

val PROJECT_SPRING_BOOT_MAP: MutableMap<String, Boolean> = mutableMapOf()
val MODULE_SPRING_BOOT_MAP: MutableMap<String, Boolean> = mutableMapOf()
val MODULE_SPRING_MAP: MutableMap<String, Boolean> = mutableMapOf()

val BINDING_VAR_MAP: MutableMap<String, PsiElement> = mutableMapOf()

fun TreeNode.findNode(o: Any): DefaultMutableTreeNode? {
    for (i in 0 until this.childCount) {
        val child = this.getChildAt(i)
        val childName = child.getNodeName()
        val oName = o.getNodeName()
        if (childName == oName) {
            return child as DefaultMutableTreeNode
        }
    }
    return null
}

fun getMethodFullName(method: PsiMethod): String {
    val builder = StringBuilder()
    builder.append(method.name).append('(')
    val paramTypes = method.typeParameterList!!.typeParameters
    val count = method.parameterList.parametersCount
    for (i in 0 until count) {
        val param = method.parameterList.getParameter(i)
        builder.append(param!!.type)
        if (i < paramTypes.size - 1) {
            builder.append(',')
        }
    }
    builder.append(')')
    return builder.toString()
}

fun resolveElement(project: Project, virtualFile: VirtualFile, element: PsiElement, func: ResolveElementFunction) {
    val fileIndex = ProjectRootManager.getInstance(project).fileIndex
    val module = fileIndex.getModuleForFile(virtualFile) ?: return
    val javaVirtualFile = ForestTemplateUtil.getSourceJavaFile(virtualFile) ?: return
    val filePath = javaVirtualFile.path
    val isTestSourceFile = ForestTemplateUtil.isTestFile(filePath)
    val hasSpringBootLib = hasSpringBootLibrary(module)
    val literal = ForestTemplateUtil.getJavaElement(element)
    val method = PsiTreeUtil.getParentOfType(literal, PsiMethod::class.java)
    func.resolve(javaVirtualFile, module, isTestSourceFile, hasSpringBootLib, method)
}

fun findMethodParameter(method: PsiMethod, func: BiFunction<PsiParameter, Int, Boolean>) {
    val paramList = PsiTreeUtil.getChildOfType(method, PsiParameterList::class.java)
    if (paramList != null && paramList.parametersCount > 0) {
        val methodParamArray = paramList.parameters
        for (i in methodParamArray.indices) {
            val methodParam = methodParamArray[i]
            if (!func.apply(methodParam, i)) {
                return
            }
        }
    }
}

fun hasSpringLibrary(module: Module): Boolean {
    val name = module.name
    if (!MODULE_SPRING_MAP.contains(name)) {
        val ret = try {
            val type = PsiType.getTypeByName(
                "org.springframework.core.SpringVersion",
                module.project,
                GlobalSearchScope.moduleScope(module)
            )
            type.isValid
        } catch (th: Throwable) {
            false
        }
        MODULE_SPRING_MAP[name] = ret
    }
    return MODULE_SPRING_MAP[name]!!
}

fun hasSpringBootLibrary(module: Module): Boolean {
    val name = module.name
    if (!MODULE_SPRING_BOOT_MAP.contains(name)) {
        val ret = try {
            val type = PsiType.getTypeByName(
                "org.springframework.boot.SpringBootVersion",
                module.project,
                GlobalSearchScope.moduleScope(module)
            )
            type.isValid
        } catch (th: Throwable) {
            false
        }
        MODULE_SPRING_BOOT_MAP[name] = ret
    }
    return MODULE_SPRING_BOOT_MAP[name]!!
}

fun hasSpringBootLibrary(project: Project): Boolean {
    val name = project.name
    if (!PROJECT_SPRING_BOOT_MAP.contains(name)) {
        val ret = try {
            val type = PsiType.getTypeByName(
                "org.springframework.boot.SpringBootVersion",
                project,
                GlobalSearchScope.allScope(project)
            )
            type.isValid
        } catch (th: Throwable) {
            false
        }
        PROJECT_SPRING_BOOT_MAP[name] = ret
    }
    return PROJECT_SPRING_BOOT_MAP[name]!!
}

fun getYAMLKeyValue(yamlFile: YAMLFile, keyName: String): Pair<YAMLPsiElement, YAMLValue?>? {
    var pair: Pair<YAMLPsiElement, YAMLValue?>? = null;
    eachYAMLKeyValues(yamlFile) { keyValue: YAMLPsiElement?, yamlValue: YAMLValue? ->
        val keyValueText = YAMLUtil.getConfigFullName(
            keyValue!!
        )
        if (keyValueText == keyName) {
            pair = Pair(keyValue, yamlValue)
            return@eachYAMLKeyValues false
        }
        true
    }
    return pair
}


fun eachYAMLKeyValues(yamlFile: YAMLFile, func: (YAMLKeyValue?, YAMLValue?) -> Boolean): Boolean {
    val topKeyValues = YAMLUtil.getTopLevelKeys(yamlFile)
    for (keyValue in topKeyValues) {
        val value = keyValue.value
        if (value != null) {
            val ret: Boolean = eachYAMLKeyValues(keyValue, value, func)
            if (!ret) {
                return false
            }
        }
    }
    return true
}

fun eachYAMLKeyValues(keyValue: YAMLKeyValue, value: YAMLValue, func: (YAMLKeyValue?, YAMLValue?) -> Boolean): Boolean {
    if (!func(keyValue, value)) {
        return false
    }
    if (value is YAMLBlockSequenceImpl) {
        val block: YAMLBlockSequenceImpl = value
        val items = block.items
        for (item in items) {
            if (item is YAMLKeyValue) {
                val ret = eachYAMLKeyValues(item, value, func)
                if (ret != null && !ret) {
                    return false
                }
            }
        }
    } else if (value is YAMLBlockMappingImpl) {
        var mapping: YAMLBlockMappingImpl = value
        var mkvs = mapping.keyValues
        for (mkv in mkvs) {
            var mvalue: YAMLValue? = mkv.value
            if (!eachYAMLKeyValues(mkv, mvalue!!, func)) {
                return false
            }
        }
    }
    return true
}

fun getPathExpressionText(element: PsiElement): String? {
    if (element is ForestTemplatePathElement) {
        val prevElementText = getPathExpressionText(element.prevSibling) ?: return element.getText()
        return prevElementText + element.getText()
    }
    return (element as? ForestTemplatePrimary)?.text
}

fun getterMethodName(name: String): String {
    return "get" + name.substring(0, 1).toUpperCase() + name.substring(1)
}

fun DefaultMutableTreeNode.findNodeOrNew(
    model: DefaultTreeModel?,
    o: Any
): DefaultMutableTreeNode {
    return this.findNodeOrNew(model, o, false)
}

fun DefaultMutableTreeNode.findNodeOrNew(
    model: DefaultTreeModel?,
    o: Any,
    sorted: Boolean
): DefaultMutableTreeNode {
    val node = this.findNode(o)
    if (node != null) {
        return node
    }
    val newNode = DefaultMutableTreeNode(o)
    if (sorted) {
        TreeUtil.insertNode(newNode, this, model, NAME_SORTED)
    } else {
        val insertionPoint = if (this.childCount >= 0) this.childCount else -(this.childCount + 1)
        if (model != null) {
            model.insertNodeInto(newNode, this, insertionPoint)
        } else {
            this.insert(newNode, insertionPoint)
        }
    }
    return newNode
}

fun DefaultMutableTreeNode.findNodeOrNew(
    model: DefaultTreeModel?,
    o: Any,
    index: Int
): DefaultMutableTreeNode {
    val node = this.findNode(o)
    return node ?: this.insertNode(model, o, index)
}

fun DefaultMutableTreeNode.moveOrInsert(
    model: DefaultTreeModel?,
    o: Any,
    index: Int
): DefaultMutableTreeNode {
    val node = this.findNode(o)
    if (node != null && this.childCount < index) {
        val child = this.getChildAt(index)
        if (child === node) {
            return node
        }
        this.removeNode(model, node)
        this.insertNode(model, node, index)
        return node
    }
    return this.insertNode(model, o, index)
}

fun DefaultMutableTreeNode.insertNode(
    model: DefaultTreeModel?,
    o: Any,
    index: Int
): DefaultMutableTreeNode {
    val newNode = DefaultMutableTreeNode(o)
    val insertionPoint = if (index >= 0) index else -(index + 1)
    if (model != null) {
        model.insertNodeInto(newNode, this, insertionPoint)
    } else {
        this.insert(newNode, insertionPoint)
    }
    return newNode
}

fun DefaultMutableTreeNode.insertNode(
    model: DefaultTreeModel?,
    child: DefaultMutableTreeNode,
    index: Int
) {
    val insertionPoint = if (index >= 0) index else -(index + 1)
    if (model != null) {
        model.insertNodeInto(child, this, insertionPoint)
    } else {
        this.insert(child, insertionPoint)
    }
}

fun DefaultMutableTreeNode.addNode(
    model: DefaultTreeModel?,
    o: Any
): DefaultMutableTreeNode {
    val newNode = DefaultMutableTreeNode(o)
    val insertionPoint = if (this.childCount >= 0) this.childCount else -(this.childCount + 1)
    if (model != null) {
        model.insertNodeInto(newNode, this, insertionPoint)
    } else {
        this.insert(newNode, insertionPoint)
    }
    return newNode
}

fun DefaultMutableTreeNode.getAllChildren(): List<MutableTreeNode> {
    val childrenList: MutableList<MutableTreeNode> = Lists.newArrayList()
    for (i in 0 until this.childCount) {
        val child = this.getChildAt(i) as MutableTreeNode
        childrenList.add(child)
    }
    return childrenList
}


fun DefaultMutableTreeNode.removeAllChildren(
    model: DefaultTreeModel?,
) {
    for (i in this.childCount - 1 downTo 0) {
        val child = this.getChildAt(i) as MutableTreeNode
        this.removeNode(model, child)
    }
}

fun DefaultMutableTreeNode.removeChildrenList(
    model: DefaultTreeModel?,
    childrenList: List<MutableTreeNode>
) {
    for (children in childrenList) {
        if (model != null) {
            model.removeNodeFromParent(children)
        } else {
            this.remove(children)
        }
    }
}

fun DefaultMutableTreeNode.removeNode(
    model: DefaultTreeModel?,
    child: MutableTreeNode?
) {
    if (model != null) {
        model.removeNodeFromParent(child)
    } else {
        this.remove(child)
    }
}

fun Any.getNodeName(): String {
    val o: Any?
    o = if (this is DefaultMutableTreeNode) {
        this.userObject
    } else {
        this
    }
    return when (o) {
        is Project -> {
            o.name
        }
        is Module -> {
            o.name
        }
        is PsiClass -> {
            o.name ?: ""
        }
        is PsiMethod -> {
            o.name
        }
        else -> {
            o?.toString() ?: "null"
        }
    }
}

fun Any.getForestAnnotationUrl(): String {
    val o: Any?
    o = if (this is DefaultMutableTreeNode) {
        this.userObject
    } else {
        this
    }
    return when (o) {
        is PsiClass -> {
            val baseRequest = o.getAnnotation(Annotation.BASE_REQUEST.qualifiedName)
            baseRequest?.let {
                // 存在@BaseRequest注解，拼接basePath
                val baseURLText = AnnotationUtil.getStringAttributeValue(it, "baseURL")
                if (!baseURLText.isNullOrBlank()) {
                    return "[${baseURLText}]"
                }
            }

            val baseURL = o.getAnnotation(Annotation.BASE_URL.qualifiedName)
            baseURL?.let {
                val valueText = AnnotationUtil.getStringAttributeValue(it, "value")
                if (!valueText.isNullOrBlank()) {
                    return "[${valueText}]"
                }
            }

            val address = o.getAnnotation(Annotation.ADDRESS.qualifiedName)
            address?.let {
                // 存在@Address注解，拼接basePath
                val schemeText = AnnotationUtil.getStringAttributeValue(it, "scheme")
                val hostText = AnnotationUtil.getStringAttributeValue(it, "host")
                val portText = AnnotationUtil.getStringAttributeValue(it, "port")
                if (!schemeText.isNullOrBlank() && !hostText.isNullOrBlank() && !portText.isNullOrBlank()) {
                    return "[$schemeText://$hostText:$portText]"
                }

                val basePathText = AnnotationUtil.getStringAttributeValue(it, "basePath")
                if (!basePathText.isNullOrBlank()) {
                    return "[${basePathText}]"
                }
            }

            ""
        }

        is PsiMethod -> {
            for ((forestMethodAnnotationName, icon) in METHOD_ICON_MAP) {
                val annotation = o.getAnnotation(forestMethodAnnotationName)
                annotation?.let {
                    val urlText = AnnotationUtil.getStringAttributeValue(it, "url")
                    if (!urlText.isNullOrBlank()) {
                        return "[${urlText}]"
                    }

                    val valueext = AnnotationUtil.getStringAttributeValue(it, "value")
                    if (!valueext.isNullOrBlank()) {
                        return "[${valueext}]"
                    }
                }

            }

            ""
        }

        else -> {
            ""
        }
    }
}

fun Any.getNodeIcon(): Icon {
    val o: Any
    o = if (this is DefaultMutableTreeNode) {
        this.userObject
    } else {
        this
    }
    return when (o) {
        is Project -> {
            AllIcons.Nodes.Folder
        }
        is Module -> {
            AllIcons.Nodes.Folder
        }
        is PsiClass -> {
            Icons.INTERFACE_18
        }
        is PsiMethod -> {
            for ((forestMethodAnnotationName, icon) in METHOD_ICON_MAP) {
                val annotation = o.getAnnotation(forestMethodAnnotationName)
                if (annotation != null) {
                    return icon
                }
            }

            val requestAnnotation = o.getAnnotation(Annotation.REQUEST.qualifiedName)
            if (requestAnnotation != null) {
                val type = AnnotationUtil.getStringAttributeValue(requestAnnotation, "type")
                val icon2 = REQUEST_TYPE_ICON_MAP[StringUtil.toUpperCase(type)]
                return icon2 ?: Icons.GET
            }

            Icons.GET
        }
        else -> {
            Icons.ICON_16
        }
    }
}

fun interface ResolveElementFunction {

    fun resolve(
        javaVirtualFile: VirtualFile,
        module: Module,
        isTestSourceFile: Boolean,
        hasSpringBootLib: Boolean,
        defMethod: PsiMethod?
    )
}
